File: cpp/n232_StackQueue.cpp
// https://leetcode-cn.com/problems/implement-queue-using-stacks/
#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>
using namespace std;

class MyQueue {
public:
    /** Initialize your data structure here. */
    MyQueue() {

    }
    
    /** Push element x to the back of queue. */
    void push(int x) {
        instack.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    int pop() {
        if (outstack.size() == 0){
            while (instack.size() != 0){
                int ele = instack.top();
                outstack.push(ele);
                instack.pop();
            }
        }
        int res = outstack.top();
        outstack.pop();
        return res;
    }
    
    /** Get the front element. */
    int peek() {
        if (outstack.size() == 0){
            while (instack.size() != 0){
                int ele = instack.top();
                outstack.push(ele);
                instack.pop();
            }
        }
        return outstack.top();
    }
    
    /** Returns whether the queue is empty. */
    bool empty() {
        return (instack.size() == 0) && (outstack.size() == 0);
    }
    
private:
    stack<int> instack;
    stack<int> outstack;
};

int main(){
    return 0;
}

File: python/n232_StackQueue.py
# https://leetcode-cn.com/problems/implement-queue-using-stacks/

class MyQueue:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.instack = []
        self.outstack = []
        

    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        self.instack.append(x)
        

    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        if len(self.outstack) == 0:
            while self.instack:
                self.outstack.append(self.instack.pop())
        return self.outstack.pop()

    def peek(self) -> int:
        """
        Get the front element.
        """
        if len(self.outstack) == 0:
            while self.instack:
                self.outstack.append(self.instack.pop())
        return self.outstack[-1]

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return len(self.instack) == 0 and len(self.outstack) == 0


if __name__ == "__main__":
    pass         